File: lib/core/utils/formatters.dart
import 'package:intl/intl.dart';
import 'constants.dart';

/// Currency and date formatting utilities
class Formatters {
  Formatters._();

  static final _currencyFormat = NumberFormat.currency(
    locale: 'en_IN',
    symbol: '', // Symbol will be prepended manually
    decimalDigits: 0,
  );

  static final _currencyFormatWithDecimals = NumberFormat.currency(
    locale: 'en_IN',
    symbol: '', // Symbol will be prepended manually
    decimalDigits: 2,
  );

  static final _compactCurrency = NumberFormat.compactCurrency(
    locale: 'en_IN',
    symbol: '', // Symbol will be prepended manually
    decimalDigits: 1,
  );

  static final _dateFormat = DateFormat('d MMM yyyy');
  static final _shortDateFormat = DateFormat('d MMM');
  static final _timeFormat = DateFormat('h:mm a');
  static final _monthFormat = DateFormat('MMMM yyyy');
  static final _dayFormat = DateFormat('EEEE');

  /// Format amount as currency: ₹1,23,456
  static String currency(double amount, {String? symbol}) {
    final s = symbol ?? AppConstants.currencySymbol;
    return '$s${_currencyFormat.format(amount).trim()}';
  }

  /// Format amount with decimals: ₹1,23,456.78
  static String currencyWithDecimals(double amount, {String? symbol}) {
    final s = symbol ?? AppConstants.currencySymbol;
    return '$s${_currencyFormatWithDecimals.format(amount).trim()}';
  }

  /// Format large amounts compactly: ₹1.2L
  static String compactCurrency(double amount, {String? symbol}) {
    final s = symbol ?? AppConstants.currencySymbol;
    return '$s${_compactCurrency.format(amount).trim()}';
  }

  /// Format as percentage: 45%
  static String percentage(double value) {
    return '${(value * 100).toStringAsFixed(0)}%';
  }

  /// Format date: 25 Jan 2026
  static String date(DateTime date) {
    return _dateFormat.format(date);
  }

  /// Format short date: 25 Jan
  static String shortDate(DateTime date) {
    return _shortDateFormat.format(date);
  }

  /// Format time: 5:30 PM
  static String time(DateTime date) {
    return _timeFormat.format(date);
  }

  /// Format date and time: 25 Jan 2026, 5:30 PM
  static String dateTime(DateTime date) {
    return '${_shortDateFormat.format(date)}, ${_timeFormat.format(date)}';
  }

  /// Format compact number: 1.5K, 2.3L
  static String compactNumber(double amount) {
    if (amount >= 100000) {
      return '${(amount / 100000).toStringAsFixed(1)}L';
    } else if (amount >= 1000) {
      return '${(amount / 1000).toStringAsFixed(1)}K';
    }
    return amount.toStringAsFixed(0);
  }

  /// Format month: January 2026
  static String month(DateTime date) {
    return _monthFormat.format(date);
  }

  /// Format day name: Sunday
  static String dayName(DateTime date) {
    return _dayFormat.format(date);
  }

  /// Format short day name: Sun
  static String dayOfWeek(DateTime date) {
    return DateFormat('E').format(date);
  }

  /// Get relative date text: Today, Yesterday, or date
  static String relativeDate(DateTime date) {
    final now = DateTime.now();
    final today = DateTime(now.year, now.month, now.day);
    final dateOnly = DateTime(date.year, date.month, date.day);

    final difference = today.difference(dateOnly).inDays;

    if (difference == 0) return 'Today';
    if (difference == 1) return 'Yesterday';
    if (difference < 7) return dayName(date);
    return shortDate(date);
  }

  /// Get days remaining in current month
  static int daysRemainingInMonth() {
    final now = DateTime.now();
    final lastDay = DateTime(now.year, now.month + 1, 0);
    return lastDay.day - now.day;
  }

  /// Get days elapsed in current month
  static int daysElapsedInMonth() {
    return DateTime.now().day;
  }

  /// Get total days in current month
  static int daysInCurrentMonth() {
    final now = DateTime.now();
    return DateTime(now.year, now.month + 1, 0).day;
  }
}

File: web_app/src/utils/formatters.js
/* Currency and date formatting utilities (exact replica of Formatters in Flutter) */
import { AppConstants } from './constants.js';

// Setup default Indian number formatting locale for Indian currency (INR) and compact formats
const inrFormatter = new Intl.NumberFormat('en-IN', {
  minimumFractionDigits: 0,
  maximumFractionDigits: 0,
});

const inrFormatterWithDecimals = new Intl.NumberFormat('en-IN', {
  minimumFractionDigits: 2,
  maximumFractionDigits: 2,
});

export const Formatters = {
  /**
   * Format amount as currency: ₹1,23,456
   */
  currency(amount, symbol) {
    const s = symbol || AppConstants.currencySymbol;
    return `${s}${inrFormatter.format(amount)}`;
  },

  /**
   * Format amount with decimals: ₹1,23,456.78
   */
  currencyWithDecimals(amount, symbol) {
    const s = symbol || AppConstants.currencySymbol;
    return `${s}${inrFormatterWithDecimals.format(amount)}`;
  },

  /**
   * Format large amounts compactly: ₹1.2L or ₹1.5K
   */
  compactCurrency(amount, symbol) {
    const s = symbol || AppConstants.currencySymbol;
    return `${s}${this.compactNumber(amount)}`;
  },

  /**
   * Format as percentage: 45%
   */
  percentage(value) {
    return `${(value * 100).toFixed(0)}%`;
  },

  /**
   * Format date: 25 Jan 2026
   */
  date(d) {
    const dateObj = new Date(d);
    return dateObj.toLocaleDateString('en-GB', {
      day: 'numeric',
      month: 'short',
      year: 'numeric'
    });
  },

  /**
   * Format short date: 25 Jan
   */
  shortDate(d) {
    const dateObj = new Date(d);
    return dateObj.toLocaleDateString('en-GB', {
      day: 'numeric',
      month: 'short'
    });
  },

  /**
   * Format time: 5:30 PM
   */
  time(d) {
    const dateObj = new Date(d);
    return dateObj.toLocaleTimeString('en-US', {
      hour: 'numeric',
      minute: '2-digit',
      hour12: true
    });
  },

  /**
   * Format date and time: 25 Jan, 5:30 PM
   */
  dateTime(d) {
    return `${this.shortDate(d)}, ${this.time(d)}`;
  },

  /**
   * Format compact number: 1.5K, 2.3L
   */
  compactNumber(amount) {
    if (amount >= 100000) {
      return `${(amount / 100000).toFixed(1)}L`;
    } else if (amount >= 1000) {
      return `${(amount / 1000).toFixed(1)}K`;
    }
    return amount.toFixed(0);
  },

  /**
   * Format month: January 2026
   */
  month(d) {
    const dateObj = new Date(d);
    return dateObj.toLocaleDateString('en-US', {
      month: 'long',
      year: 'numeric'
    });
  },

  /**
   * Format day name: Sunday
   */
  dayName(d) {
    const dateObj = new Date(d);
    return dateObj.toLocaleDateString('en-US', { weekday: 'long' });
  },

  /**
   * Format short day name: Sun
   */
  dayOfWeek(d) {
    const dateObj = new Date(d);
    return dateObj.toLocaleDateString('en-US', { weekday: 'short' });
  },

  /**
   * Get relative date text: Today, Yesterday, or date
   */
  relativeDate(d) {
    const dateVal = new Date(d);
    const now = new Date();
    
    const today = new Date(now.getFullYear(), now.getMonth(), now.getDate());
    const targetDate = new Date(dateVal.getFullYear(), dateVal.getMonth(), dateVal.getDate());
    
    const diffTime = today - targetDate;
    const diffDays = Math.round(diffTime / (1000 * 60 * 60 * 24));
    
    if (diffDays === 0) return 'Today';
    if (diffDays === 1) return 'Yesterday';
    if (diffDays < 7 && diffDays > 0) return this.dayName(d);
    return this.shortDate(d);
  },

  /**
   * Get days remaining in current month
   */
  daysRemainingInMonth() {
    const now = new Date();
    const lastDay = new Date(now.getFullYear(), now.getMonth() + 1, 0);
    return lastDay.getDate() - now.getDate();
  },

  /**
   * Get days elapsed in current month
   */
  daysElapsedInMonth() {
    return new Date().getDate();
  },

  /**
   * Get total days in current month
   */
  daysInCurrentMonth() {
    const now = new Date();
    return new Date(now.getFullYear(), now.getMonth() + 1, 0).getDate();
  }
};
